trizen.1
.TH TRIZEN "1" "February 2022" "trizen 1.68" "User Commands"

.SH NAME
trizen \- a lightweight wrapper for AUR, written in Perl.

.SH "SYNOPSIS"
.sp
\fItrizen\fR <operation> [options] [targets]

.SH "OPERATIONS"
.PP
\fB\-S, \-\-sync\fR
.RS 4
Synchronize packages. Packages are installed directly from the remote repositories or from the AUR, including all dependencies required to run the packages.
.RE
.PP
\fB\-C, \-\-comments\fR
.RS 4
Display AUR comments for a package.
.RE
.PP
\fB\-G, \-\-get\fR
.RS 4
Clone an AUR package in the current directory.
.RE
.PP
\fB\-R, \-\-remove\fR
.RS 4
Remove package(s) from the system. See: \fIpacman -Rh\fR
.RE
.PP
\fB\-Q, \-\-query\fR
.RS 4
Query the package database. See: \fIpacman -Qh\fR
.RE
.PP
\fB\-F, \-\-files\fR
.RS 4
Query the files database. See: \fIpacman -Fh\fR
.RE
.PP
\fB\-D, \-\-database\fR
.RS 4
Operate on the package database. See: \fIpacman -Dh\fR
.RE
.PP
\fB\-T, \-\-deptest\fR
.RS 4
Check dependencies. See: \fIpacman -Th\fR
.RE
.PP
\fB\-U, \-\-upgrade\fR
.RS 4
Install built packages from \fI--clone-dir\fR or \fI`pwd`\fR. See: \fIpacman -Uh\fR

.RE
.SH "OPTIONS"
.PP
\fB\-q, \-\-quiet\fR
.RS 4
Display minimal output.
.RE
.PP
\fB\-a, \-\-aur\fR
.RS 4
Perform only AUR operations.
.RE
.PP
\fB\-r, \-\-regular\fR
.RS 4
Perform only regular \fIpacman\fR operations (the opposite of \fI--aur\fR).
.RE
.PP
\fB\-\-stats\fR
.RS 4
Display statistics for the installed packages.
.RE
.PP
\fB\-\-nocolors\fR
.RS 4
Disable ANSI colors.
.RE
.PP
\fB\-\-forcecolors\fR
.RS 4
Enable colors even when not writing to STDOUT.
.RE
.PP
\fB\-\-debug\fR
.RS 4
Enable verbose mode.

.RE
.SH "SYNC OPTIONS (APPLY TO \FI-S\FR)"
.PP
\fB\-s, \-\-search\fR
.RS 4
Search for packages. By default, it includes packages from the pacman repos and from the AUR.
.RE
.PP
\fB\-i, \-\-info\fR
.RS 4
Display info for packages.
.RE
.PP
\fB\-m, \-\-maintainer\fR
.RS 4
Display the AUR packages maintained by a given \fIusername\fR.
.RE
.PP
\fB\-p, \-\-pkgbuild\fR
.RS 4
Display the PKGBUILD of a given AUR package. For repo packages, it displays the download link.
.RE
.PP
\fB\-l, \-\-local\fR
.RS 4
Build and install packages from the current directory. This option is best used in combination with \fI-G\fR.

After a package is cloned in the current directory, the user can inspect or edit the build files, then \fI-Sl <pkgname>\fR will build and install the package, without pulling new changes from the AUR.
.RE
.PP
\fB\-u, \-\-sysupgrade\fR
.RS 4
Upgrades all packages that are out-of-date.
.RE
.PP
\fB\-y, \-\-refresh\fR
.RS 4
Refresh package databases.
.RE
.PP
\fB\-c, \-\-clean\fR
.RS 4
Clean the cache directories of \fIpacman\fR and \fItrizen\fR. Add \fI-r\fR to clean only pacman's cache, or \fI-a\fR to clean only trizen's cache.
.RE
.PP
\fB\-\-devel\fR
.RS 4
Update VCS packages during \fI-Su\fR.

In combination with \fI--needed\fR, trizen will check if a package really needs to be updated (i.e. has a new commit). Currently, only git sources are supported.
.RE
.PP
\fB\-\-show-ood\fR
.RS 4
Show out-of-date flagged packages during \fI-Su\fR.
.RE
.PP
\fB\-\-noinfo\fR
.RS 4
Do not display package information after cloning.
.RE
.PP
\fB\-\-nopull\fR
.RS 4
Do not \fI`git pull`\fR new changes from the AUR when the build files of a package already exist locally.
.RE
.PP
\fB\-\-noedit\fR
.RS 4
Do not prompt to edit the build files of packages.
.RE
.PP
\fB\-\-nobuild\fR
.RS 4
Do not build packages (implies \fI--noedit\fR). Assumes a built package already exists.
.RE
.PP
\fB\-\-noinstall\fR
.RS 4
Do not install packages after building.
.RE
.PP
\fB\-\-nocheck\fR
.RS 4
Do not check() packages or install check dependencies.
.RE
.PP
\fB\-\-noconfirm\fR
.RS 4
Do not ask for any confirmation.
.RE
.PP
\fB\-\-needed\fR
.RS 4
Do not reinstall up-to-date packages.
.RE
.PP
\fB\-\-asdeps\fR
.RS 4
Install packages as dependencies (non-explicit).
.RE
.PP
\fB\-\-asexplicit\fR
.RS 4
Install packages as explicitly installed.
.RE
.PP
\fB\-\-skipinteg\fR
.RS 4
Pass the \fI--skipinteg\fR argument to \fImakepkg\fR.
.RE
.PP
\fB\-\-movepkg\fR
.RS 4
Move built packages in pacman's cache directory.
.RE
.PP
\fB\-\-movepkg-dir='...'\fR
.RS 4
Move built packages into this directory (implies \fI--movepkg\fR).
.RE
.PP
\fB\-\-clone-dir='...'\fR
.RS 4
Directory where to clone and build packages.
.RE
.PP
\fB\-\-editor='...'\fR
.RS 4
Editor command used to edit the build files of AUR packages. The default editor is \fI$VISUAL\fR or \fI$EDITOR\fR (in this order).
.RE
.PP
\fB\-\-pager-mode\fR
.RS 4
Display the build files of packages in pager mode.
.RE
.PP
\fB\-\-pager='...'\fR
.RS 4
Pager command used to display the build files of packages (with \fI--pager-mode\fR). The default pager is \fI$PAGER\fR.
.RE
.PP
\fB\-\-ignore='...'\fR
.RS 4
Space-separated list of packages to ignore during \fI-Su\fR.
.RE

.RE
.SH "QUERY AUR OPTIONS"
.PP
\fB\-Qua\fR
.RS 4
Display AUR updates only.
.RE
.PP
\fB\-Qma\fR
.RS 4
Display foreign installed packages that do not exist in repos or in the AUR.
.RE

.RE
.SH "GET OPTIONS (APPLY TO \FI-G\FR)"
.PP
\fB\-d, \-\-with\-deps\fR
.RS 4
Clone a package along with all its AUR dependencies that are not already installed.
.RE

.SH "INTERACTIVE MODE"
.PP
Search and select packages to install, using the following syntax:

.RS 4
\fItrizen [keywords]\fR
.RE
.PP
By default, the results include both repo and AUR packages. Add \fI--regular\fR or \fI--aur\fR to search only for repo or AUR packages, respectively.

.SH "EXAMPLES"
.nf
\ \ trizen\ \-S\ \ <package>\ \ \ \ \ #\ install\ <package>
\ \ trizen\ \-Ss\ <keyword>\ \ \ \ \ #\ search\ for\ <keyword>
\ \ trizen\ \-Si\ <package>\ \ \ \ \ #\ show\ info\ about\ <package>

\ \ trizen\ \-G\ \ <package>\ \ \ \ \ #\ clones\ <package>
\ \ trizen\ \-Gd\ <package>\ \ \ \ \ #\ clones\ <package>\ along\ with\ its\ AUR\ dependencies

.SH CONFIGURATION
Configuration file: ~/.config/trizen/trizen.conf

.SH "CONFIGURATION OPTIONS"
.PP
Each configuration key can be used as a command-line argument, by preceding it with `\fI--\fR` and (optionally) replacing underscores (`\fI_\fR`) with dashes (`\fI-\fR`).

Example:

.RS 4
\fItrizen --stats --packages-in-stats=50\fR
.RE
.PP
The above command will display 50 packages in \fI--stats\fR.

.SH AUTHOR
Daniel Șuteu <trizen@protonmail.com>

.SH "SEE ALSO"
Upstream repository: https://github.com/trizen/trizen


ansible/trizen.py
#!/usr/bin/env python

# The MIT License (MIT)
#
# Copyright (c) 2017 Plague Doctor
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software, to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish, distribute,
# sub-license, and/or sell copies of the Software, and to permit persons to whom
# the Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
# This module is a re-written version of the pacaur.py module by Austin Hyde.


from ansible.module_utils.basic import *


def trizen_available(module):
    rc, stdout, stderr = module.run_command('which trizen', check_rc=False)
    return rc == 0


def pacman_available(module):
    rc, stdout, stderr = module.run_command('which pacman', check_rc=False)
    return rc == 0


def package_installed(module, pkg):
    rc, stdout, stderr = module.run_command('pacman -Q {}'.format(pkg), check_rc=False)
    return rc == 0


def check_packages(module, pkgs, state):
    global packages
    would_be_changed = []
    action = 'installed'

    for pkg in packages:
        installed = package_installed(module, pkg)
        if (state == 'present' and not installed) or (state == 'absent' and installed):
            would_be_changed.append(pkg)

    if state == 'absent':
        action = 'removed'

    if would_be_changed:
        module.exit_json(changed=True, msg='{no_of_packages} package(s) would be {act}.'.format(no_of_packages=len(would_be_changed), act=action))
    else:
        module.exit_json(changed=False, msg='all packages are already {act}.'.format(act=action))


def install_packages(module, pkgs):
    num_installed = 0

    cmd = 'trizen --quiet --noconfirm -S {pkg_name}'

    for pkg in pkgs:
        if package_installed(module, pkg):
            continue

        rc, stdout, stderr = module.run_command(cmd.format(pkg_name=pkg), check_rc=False)

        if rc != 0:
            module.fail_json(msg='failed to install package {pkg_name}, because: {err}.'.format(pkg_name=pkg, err=stderr))

        num_installed += 1

    if num_installed > 0:
        module.exit_json(changed=True, msg='installed {} package(s).'.format(num_installed))
    else:
        module.exit_json(changed=False, msg='all packages were already installed.')


def remove_packages(module, pkgs, recurse):
    num_removed = 0

    arg = 'R'
    action = 'remove'
    if recurse:
        arg = 'Rs'
        action = 'recursively remove'

    cmd = 'pacman -{argument} --noconfirm {pkg_name}'

    for pkg in pkgs:
        if not package_installed(module, pkg):
            continue

        rc, stdout, stderr = module.run_command(cmd.format(argument=arg, pkg_name=pkg), check_rc=False)

        if rc != 0:
            module.fail_json(msg='failed to {act} package {pkg_name} because: {err}.'.format(act=action, pkg_name=pkg, err=stderr))

        num_removed += 1

    if num_removed > 0:
        module.exit_json(changed=True, msg='removed {no_of_packages} package(s).'.format(no_of_packages=num_removed))
    else:
        module.exit_json(changed=False, msg='all packages were already removed.')


def main():
    module = AnsibleModule(
        argument_spec=dict(
            name=dict(required=True),
            state=dict(default='present', choices=['present', 'absent']),
            recurse=dict(default='no', choices=BOOLEANS, type='bool')
        ),
        supports_check_mode=True
    )

    if not trizen_available(module):
        module.fail_json(msg="could not locate trizen...")

    if not pacman_available(module):
        module.fail_json(msg="could not locate pacman...")

    p = module.params

    pkgs = p['name'].split(',')

    if module.check_mode:
        check_packages(module, pkgs, p['state'])

    if p['state'] == 'present':
        install_packages(module, pkgs)
    elif p['state'] == 'absent':
        remove_packages(module, pkgs, p['recurse'])


if __name__ == '__main__':
    main()
